File: _legacy/packages/core/src/ask/ask.test.ts
import { describe, it, expect, vi, beforeEach, afterEach } from 'vitest';
import { askQuestion, buildAskPrompt } from './ask.js';
import type { SearchResult } from '../search/query.js';

// Mock dependencies
vi.mock('../config/config.js', () => ({
  loadConfig: vi.fn(),
  loadConfigAsync: vi.fn(),
  isConfigured: vi.fn(),
}));

vi.mock('../search/query.js', () => ({
  searchRepos: vi.fn(),
}));

vi.mock('../gemini/llm.js', () => ({
  streamLLM: vi.fn(),
  isLLMError: vi.fn((result) => 'error' in result && !('text' in result)),
}));

// Import mocked modules
import { loadConfig, loadConfigAsync, isConfigured } from '../config/config.js';
import { searchRepos } from '../search/query.js';
import { streamLLM } from '../gemini/llm.js';

const mockedLoadConfig = vi.mocked(loadConfig);
const mockedLoadConfigAsync = vi.mocked(loadConfigAsync);
const mockedIsConfigured = vi.mocked(isConfigured);
const mockedSearchRepos = vi.mocked(searchRepos);
const mockedStreamLLM = vi.mocked(streamLLM);

/**
 * Helper to create a mock SearchResult.
 */
function createMockSearchResult(
  repoFullName: string,
  chunkType: 'metadata' | 'readme' | 'file_tree',
  similarity: number,
  content: string = 'Test content'
): SearchResult {
  const [owner, name] = repoFullName.split('/');
  return {
    repoFullName,
    owner,
    repoName: name,
    description: 'Test description',
    language: 'TypeScript',
    stars: 100,
    isStarred: true,
    isOwned: false,
    htmlUrl: `https://github.com/${repoFullName}`,
    chunkType,
    content,
    similarity,
  };
}

describe('ask/ask', () => {
  beforeEach(() => {
    vi.clearAllMocks();

    // Default mocks
    mockedIsConfigured.mockReturnValue(true);
    mockedLoadConfig.mockReturnValue({
      githubPat: null,
      geminiKey: null,
      dbPath: '/test/db.db',
    });
    mockedLoadConfigAsync.mockResolvedValue({
      githubPat: 'test-pat',
      geminiKey: 'test-gemini-key',
      dbPath: '/test/db.db',
    });
  });

  afterEach(() => {
    vi.clearAllMocks();
  });

  describe('askQuestion', () => {
    it('returns a populated AskResult with answer and sources on success', async () => {
      const mockResults: SearchResult[] = [
        createMockSearchResult('owner/repo1', 'readme', 0.95, 'README content'),
        createMockSearchResult('owner/repo2', 'metadata', 0.85, 'Metadata content'),
      ];

      mockedSearchRepos.mockResolvedValue({
        results: mockResults,
        totalConsidered: 2,
        queryEmbeddingMs: 50,
        searchMs: 10,
      });

      mockedStreamLLM.mockResolvedValue({
        text: 'This is the answer based on the context.',
        inputTokens: 100,
        outputTokens: 50,
        durationMs: 500,
      });

      const result = await askQuestion({ question: 'What is this project?' });

      expect(result.answer).toBe('This is the answer based on the context.');
      expect(result.question).toBe('What is this project?');
      expect(result.sources).toHaveLength(2);
      expect(result.sources[0].repoFullName).toBe('owner/repo1');
      expect(result.sources[0].similarity).toBe(0.95);
      expect(result.inputTokens).toBe(100);
      expect(result.outputTokens).toBe(50);
      expect(result.durationMs).toBeGreaterThanOrEqual(0);
    });

    it('calls onChunk for each streamed chunk during askQuestion', async () => {
      const mockResults: SearchResult[] = [
        createMockSearchResult('owner/repo1', 'readme', 0.9),
      ];

      mockedSearchRepos.mockResolvedValue({
        results: mockResults,
        totalConsidered: 1,
        queryEmbeddingMs: 50,
        searchMs: 10,
      });

      // Capture the onChunk callback and simulate calling it
      mockedStreamLLM.mockImplementation(async (_apiKey, _prompt, _systemPrompt, onChunk) => {
        if (onChunk) {
          onChunk('First ');
          onChunk('chunk.');
        }
        return {
          text: 'First chunk.',
          inputTokens: 50,
          outputTokens: 20,
          durationMs: 300,
        };
      });

      const chunks: string[] = [];
      const onChunk = (text: string): void => {
        chunks.push(text);
      };

      await askQuestion({ question: 'Test question' }, onChunk);

      expect(chunks).toHaveLength(2);
      expect(chunks[0]).toBe('First ');
      expect(chunks[1]).toBe('chunk.');
    });

    it('filters results to only specified repo when options.repo is provided', async () => {
      const mockResults: SearchResult[] = [
        createMockSearchResult('owner/repo1', 'readme', 0.95),
        createMockSearchResult('owner/repo2', 'metadata', 0.85),
        createMockSearchResult('other/repo3', 'file_tree', 0.80),
      ];

      mockedSearchRepos.mockResolvedValue({
        results: mockResults,
        totalConsidered: 3,
        queryEmbeddingMs: 50,
        searchMs: 10,
      });

      mockedStreamLLM.mockResolvedValue({
        text: 'Answer about repo1.',
        inputTokens: 80,
        outputTokens: 30,
        durationMs: 400,
      });

      const result = await askQuestion({
        question: 'What is repo1?',
        repo: 'owner/repo1',
      });

      // Verify streamLLM was called - the prompt should only include repo1
      expect(mockedStreamLLM).toHaveBeenCalled();
      const [, prompt] = mockedStreamLLM.mock.calls[0];
      expect(prompt).toContain('owner/repo1');
      expect(prompt).not.toContain('owner/repo2');
      expect(prompt).not.toContain('other/repo3');

      // Sources should only contain repo1
      expect(result.sources).toHaveLength(1);
      expect(result.sources[0].repoFullName).toBe('owner/repo1');
    });

    it('returns safe AskResult with no-information message when searchRepos returns empty', async () => {
      mockedSearchRepos.mockResolvedValue({
        results: [],
        totalConsidered: 0,
        queryEmbeddingMs: 50,
        searchMs: 10,
      });

      const result = await askQuestion({ question: 'Unknown topic?' });

      expect(result.answer).toBe(
        "I couldn't find any relevant information in your knowledge base to answer this question."
      );
      expect(result.sources).toHaveLength(0);
      expect(result.question).toBe('Unknown topic?');
      // Should not throw
      expect(result.inputTokens).toBe(0);
      expect(result.outputTokens).toBe(0);
    });

    it('returns safe AskResult with error message when streamLLM returns LLMError', async () => {
      const mockResults: SearchResult[] = [
        createMockSearchResult('owner/repo1', 'readme', 0.9),
      ];

      mockedSearchRepos.mockResolvedValue({
        results: mockResults,
        totalConsidered: 1,
        queryEmbeddingMs: 50,
        searchMs: 10,
      });

      mockedStreamLLM.mockResolvedValue({
        error: 'API rate limit exceeded',
        durationMs: 100,
      });

      const result = await askQuestion({ question: 'Test question' });

      expect(result.answer).toContain('Error generating answer');
      expect(result.answer).toContain('API rate limit exceeded');
      expect(result.sources).toHaveLength(0);
      // Should not throw
      expect(result.question).toBe('Test question');
    });

    it('deduplicates sources by repoFullName - only one entry per repo', async () => {
      const mockResults: SearchResult[] = [
        createMockSearchResult('owner/repo1', 'readme', 0.95),
        createMockSearchResult('owner/repo1', 'metadata', 0.90), // Same repo, different chunk
        createMockSearchResult('owner/repo1', 'file_tree', 0.85), // Same repo again
        createMockSearchResult('owner/repo2', 'readme', 0.80),
      ];

      mockedSearchRepos.mockResolvedValue({
        results: mockResults,
        totalConsidered: 4,
        queryEmbeddingMs: 50,
        searchMs: 10,
      });

      mockedStreamLLM.mockResolvedValue({
        text: 'Answer.',
        inputTokens: 100,
        outputTokens: 10,
        durationMs: 200,
      });

      const result = await askQuestion({ question: 'Test' });

      // Should have only 2 unique repos
      expect(result.sources).toHaveLength(2);
      const repoNames = result.sources.map((s) => s.repoFullName);
      expect(repoNames).toContain('owner/repo1');
      expect(repoNames).toContain('owner/repo2');
      // Should keep highest similarity for repo1 (0.95, not 0.90 or 0.85)
      const repo1Source = result.sources.find((s) => s.repoFullName === 'owner/repo1');
      expect(repo1Source?.similarity).toBe(0.95);
    });

    it('sorts sources by similarity descending and caps at 5', async () => {
      const mockResults: SearchResult[] = [
        createMockSearchResult('owner/repo1', 'readme', 0.60),
        createMockSearchResult('owner/repo2', 'readme', 0.95),
        createMockSearchResult('owner/repo3', 'readme', 0.70),
        createMockSearchResult('owner/repo4', 'readme', 0.80),
        createMockSearchResult('owner/repo5', 'readme', 0.85),
        createMockSearchResult('owner/repo6', 'readme', 0.50), // Should be excluded (6th)
        createMockSearchResult('owner/repo7', 'readme', 0.40), // Should be excluded (7th)
      ];

      mockedSearchRepos.mockResolvedValue({
        results: mockResults,
        totalConsidered: 7,
        queryEmbeddingMs: 50,
        searchMs: 10,
      });

      mockedStreamLLM.mockResolvedValue({
        text: 'Answer.',
        inputTokens: 150,
        outputTokens: 10,
        durationMs: 200,
      });

      const result = await askQuestion({ question: 'Test' });

      // Should be capped at 5
      expect(result.sources).toHaveLength(5);

      // Should be sorted by similarity descending
      expect(result.sources[0].similarity).toBe(0.95);
      expect(result.sources[1].similarity).toBe(0.85);
      expect(result.sources[2].similarity).toBe(0.80);
      expect(result.sources[3].similarity).toBe(0.70);
      expect(result.sources[4].similarity).toBe(0.60);

      // repo6 and repo7 should be excluded
      const repoNames = result.sources.map((s) => s.repoFullName);
      expect(repoNames).not.toContain('owner/repo6');
      expect(repoNames).not.toContain('owner/repo7');
    });

    it('durationMs is present and non-negative', async () => {
      mockedSearchRepos.mockResolvedValue({
        results: [],
        totalConsidered: 0,
        queryEmbeddingMs: 50,
        searchMs: 10,
      });

      const result = await askQuestion({ question: 'Test' });

      expect(typeof result.durationMs).toBe('number');
      expect(result.durationMs).toBeGreaterThanOrEqual(0);
    });
  });

  describe('buildAskPrompt', () => {
    it('includes all chunk contents in the formatted context block', () => {
      const chunks: SearchResult[] = [
        createMockSearchResult('owner/repo1', 'readme', 0.95, 'README content here'),
        createMockSearchResult('owner/repo2', 'metadata', 0.85, 'Metadata content here'),
        createMockSearchResult('owner/repo3', 'file_tree', 0.70, 'File tree content here'),
      ];

      const prompt = buildAskPrompt('What are these projects?', chunks);

      // Check question is included
      expect(prompt).toContain('Question: What are these projects?');

      // Check all chunks are formatted correctly
      expect(prompt).toContain('--- owner/repo1 (readme) ---');
      expect(prompt).toContain('README content here');

      expect(prompt).toContain('--- owner/repo2 (metadata) ---');
      expect(prompt).toContain('Metadata content here');

      expect(prompt).toContain('--- owner/repo3 (file_tree) ---');
      expect(prompt).toContain('File tree content here');

      // Check instructions are included
      expect(prompt).toContain('Answer the question based on the context above');
    });
  });
});

File: internal/ask/ask_test.go
package ask

import (
	"context"
	"database/sql"
	"path/filepath"
	"testing"

	"github.com/hackastak/repog/internal/db"
	"github.com/hackastak/repog/internal/provider"
	"github.com/hackastak/repog/internal/search"
)

func TestBuildSourceAttributions(t *testing.T) {
	results := []search.SearchResult{
		{RepoFullName: "user/repo1", ChunkType: "readme", Similarity: 0.9},
		{RepoFullName: "user/repo1", ChunkType: "metadata", Similarity: 0.8},
		{RepoFullName: "user/repo2", ChunkType: "readme", Similarity: 0.7},
	}

	sources := buildSourceAttributions(results)

	// Should deduplicate by repo, keeping highest similarity
	if len(sources) != 2 {
		t.Errorf("Expected 2 sources, got %d", len(sources))
	}

	// Check that we keep the highest similarity for deduped repos
	for _, s := range sources {
		if s.RepoFullName == "user/repo1" && s.Similarity != 0.9 {
			t.Errorf("Expected highest similarity 0.9 for user/repo1, got %f", s.Similarity)
		}
	}
}

func TestBuildAskPrompt(t *testing.T) {
	chunks := []search.SearchResult{
		{RepoFullName: "user/repo1", ChunkType: "readme", Content: "This is the README content"},
	}

	prompt := buildAskPrompt("What is this repo about?", chunks)

	if !containsString(prompt, "What is this repo about?") {
		t.Error("Prompt should contain the question")
	}

	if !containsString(prompt, "user/repo1") {
		t.Error("Prompt should contain the repo name")
	}

	if !containsString(prompt, "This is the README content") {
		t.Error("Prompt should contain the chunk content")
	}
}

func containsString(s, substr string) bool {
	return len(s) >= len(substr) && (s == substr || (len(s) > 0 && containsHelper(s, substr)))
}

func containsHelper(s, substr string) bool {
	for i := 0; i <= len(s)-len(substr); i++ {
		if s[i:i+len(substr)] == substr {
			return true
		}
	}
	return false
}

func insertTestRepoWithEmbedding(t *testing.T, database *sql.DB, fullName string, content string, embedding []float32) int64 {
	t.Helper()

	result, err := database.Exec(`
		INSERT INTO repos (github_id, full_name, owner, name, description, stars, language, topics, html_url, synced_at)
		VALUES (?, ?, 'owner', 'name', 'desc', 100, 'Go', '[]', 'https://github.com/test', datetime('now'))
	`, 12345, fullName)
	if err != nil {
		t.Fatalf("Failed to insert repo: %v", err)
	}

	repoID, err := result.LastInsertId()
	if err != nil {
		t.Fatalf("Failed to get repo ID: %v", err)
	}

	chunkResult, err := database.Exec(`
		INSERT INTO chunks (repo_id, chunk_type, content)
		VALUES (?, 'readme', ?)
	`, repoID, content)
	if err != nil {
		t.Fatalf("Failed to insert chunk: %v", err)
	}

	chunkID, err := chunkResult.LastInsertId()
	if err != nil {
		t.Fatalf("Failed to get chunk ID: %v", err)
	}

	embeddingBlob := provider.Float32SliceToBytes(embedding)
	_, err = database.Exec("INSERT INTO chunk_embeddings (chunk_id, embedding) VALUES (?, ?)", chunkID, embeddingBlob)
	if err != nil {
		t.Fatalf("Failed to insert embedding: %v", err)
	}

	return repoID
}

func TestAskQuestion_ReturnsPopulatedResult(t *testing.T) {
	// Create mock providers
	mockEmbed := provider.NewMockEmbeddingProvider()
	mockEmbed.QueryFunc = func(_ context.Context, _ string) []float32 {
		embedding := make([]float32, 768)
		for i := range embedding {
			embedding[i] = 0.5
		}
		return embedding
	}

	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, onChunk func(string)) (*provider.LLMResult, *provider.LLMError) {
			text := "This is the answer based on the context."
			if onChunk != nil {
				onChunk(text)
			}
			return &provider.LLMResult{Text: text, InputTokens: 100, OutputTokens: 50}, nil
		},
	}

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	// Create embedding that will match query embedding (0.5 values)
	embedding := make([]float32, 768)
	for i := range embedding {
		embedding[i] = 0.5
	}
	insertTestRepoWithEmbedding(t, database, "user/testrepo", "This is repo content.", embedding)

	var receivedChunks []string
	onChunk := func(chunk string) {
		receivedChunks = append(receivedChunks, chunk)
	}

	result, err := AskQuestion(context.Background(), AskOptions{
		Question:          "What is this repo about?",
		DB:                database,
		EmbeddingProvider: mockEmbed,
		LLMProvider:       mockLLM,
		Limit:             10,
	}, onChunk)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.Question != "What is this repo about?" {
		t.Errorf("Expected question preserved, got '%s'", result.Question)
	}

	if result.Answer == "" {
		t.Error("Expected non-empty answer")
	}

	if len(receivedChunks) == 0 {
		t.Error("Expected onChunk to be called")
	}

	if result.DurationMs < 0 {
		t.Errorf("Expected non-negative duration, got %d", result.DurationMs)
	}

	if len(result.Sources) == 0 {
		t.Error("Expected sources to be populated")
	}
}

func TestAskQuestion_RepoFilterScopesResults(t *testing.T) {
	mockEmbed := provider.NewMockEmbeddingProvider()
	mockEmbed.QueryFunc = func(_ context.Context, _ string) []float32 {
		embedding := make([]float32, 768)
		for i := range embedding {
			embedding[i] = 0.5
		}
		return embedding
	}

	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, onChunk func(string)) (*provider.LLMResult, *provider.LLMError) {
			text := "Filtered answer"
			if onChunk != nil {
				onChunk(text)
			}
			return &provider.LLMResult{Text: text, InputTokens: 10, OutputTokens: 5}, nil
		},
	}

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	// Create embeddings
	embedding := make([]float32, 768)
	for i := range embedding {
		embedding[i] = 0.5
	}

	// Insert two repos with unique github_id
	result1, _ := database.Exec(`
		INSERT INTO repos (github_id, full_name, owner, name, description, stars, language, topics, html_url, synced_at)
		VALUES (1001, 'user/repo1', 'user', 'repo1', 'desc', 100, 'Go', '[]', 'https://github.com/user/repo1', datetime('now'))
	`)
	repo1ID, _ := result1.LastInsertId()
	chunk1Result, _ := database.Exec(`INSERT INTO chunks (repo_id, chunk_type, content) VALUES (?, 'readme', 'Repo1 content')`, repo1ID)
	chunk1ID, _ := chunk1Result.LastInsertId()
	_, _ = database.Exec("INSERT INTO chunk_embeddings (chunk_id, embedding) VALUES (?, ?)", chunk1ID, provider.Float32SliceToBytes(embedding))

	result2, _ := database.Exec(`
		INSERT INTO repos (github_id, full_name, owner, name, description, stars, language, topics, html_url, synced_at)
		VALUES (1002, 'user/repo2', 'user', 'repo2', 'desc', 100, 'Go', '[]', 'https://github.com/user/repo2', datetime('now'))
	`)
	repo2ID, _ := result2.LastInsertId()
	chunk2Result, _ := database.Exec(`INSERT INTO chunks (repo_id, chunk_type, content) VALUES (?, 'readme', 'Repo2 content')`, repo2ID)
	chunk2ID, _ := chunk2Result.LastInsertId()
	_, _ = database.Exec("INSERT INTO chunk_embeddings (chunk_id, embedding) VALUES (?, ?)", chunk2ID, provider.Float32SliceToBytes(embedding))

	// Filter to only user/repo1
	askResult, err := AskQuestion(context.Background(), AskOptions{
		Question:          "What is this repo about?",
		Repo:              "user/repo1",
		DB:                database,
		EmbeddingProvider: mockEmbed,
		LLMProvider:       mockLLM,
		Limit:             10,
	}, nil)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	// All sources should be from user/repo1
	for _, source := range askResult.Sources {
		if source.RepoFullName != "user/repo1" {
			t.Errorf("Expected source from user/repo1, got %s", source.RepoFullName)
		}
	}
}

func TestAskQuestion_ReturnsNoInfoMessageWhenNoResults(t *testing.T) {
	mockEmbed := provider.NewMockEmbeddingProvider()
	mockEmbed.QueryFunc = func(_ context.Context, _ string) []float32 {
		embedding := make([]float32, 768)
		return embedding
	}

	// LLM should not be called when no results
	mockLLM := provider.NewMockLLMProvider()

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	// Empty database - no repos

	var receivedChunk string
	onChunk := func(chunk string) {
		receivedChunk = chunk
	}

	result, err := AskQuestion(context.Background(), AskOptions{
		Question:          "What is this repo about?",
		DB:                database,
		EmbeddingProvider: mockEmbed,
		LLMProvider:       mockLLM,
	}, onChunk)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	expectedMsg := "I couldn't find any relevant information in your knowledge base to answer this question."
	if result.Answer != expectedMsg {
		t.Errorf("Expected no info message, got '%s'", result.Answer)
	}

	if receivedChunk != expectedMsg {
		t.Errorf("Expected onChunk to receive no info message, got '%s'", receivedChunk)
	}
}

func TestAskQuestion_HandlesLLMError(t *testing.T) {
	mockEmbed := provider.NewMockEmbeddingProvider()
	mockEmbed.QueryFunc = func(_ context.Context, _ string) []float32 {
		embedding := make([]float32, 768)
		for i := range embedding {
			embedding[i] = 0.5
		}
		return embedding
	}

	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, _ func(string)) (*provider.LLMResult, *provider.LLMError) {
			return nil, &provider.LLMError{Message: "Internal Server Error", StatusCode: 500}
		},
	}

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	embedding := make([]float32, 768)
	for i := range embedding {
		embedding[i] = 0.5
	}
	insertTestRepoWithEmbedding(t, database, "user/repo", "Content", embedding)

	result, err := AskQuestion(context.Background(), AskOptions{
		Question:          "What is this repo about?",
		DB:                database,
		EmbeddingProvider: mockEmbed,
		LLMProvider:       mockLLM,
	}, nil)

	// Should not return Go error, but Answer should contain error message
	if err != nil {
		t.Fatalf("Expected no Go error, got %v", err)
	}

	if !containsString(result.Answer, "Error generating answer") {
		t.Errorf("Expected error message in answer, got '%s'", result.Answer)
	}
}

func TestAskQuestion_DefaultLimit(t *testing.T) {
	mockEmbed := provider.NewMockEmbeddingProvider()
	mockEmbed.QueryFunc = func(_ context.Context, _ string) []float32 {
		embedding := make([]float32, 768)
		return embedding
	}

	mockLLM := provider.NewMockLLMProvider()

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	// Test with Limit = 0 (should default to 10)
	result, err := AskQuestion(context.Background(), AskOptions{
		Question:          "What is this repo about?",
		DB:                database,
		EmbeddingProvider: mockEmbed,
		LLMProvider:       mockLLM,
		Limit:             0, // Should default to 10
	}, nil)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	// Since there's no data, we just verify no panic occurred
	if result.Question != "What is this repo about?" {
		t.Errorf("Expected question preserved, got '%s'", result.Question)
	}
}

func TestAskQuestion_NilOnChunkCallback(t *testing.T) {
	mockEmbed := provider.NewMockEmbeddingProvider()
	mockEmbed.QueryFunc = func(_ context.Context, _ string) []float32 {
		embedding := make([]float32, 768)
		for i := range embedding {
			embedding[i] = 0.5
		}
		return embedding
	}

	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, onChunk func(string)) (*provider.LLMResult, *provider.LLMError) {
			if onChunk != nil {
				onChunk("Answer")
			}
			return &provider.LLMResult{Text: "Answer", InputTokens: 10, OutputTokens: 5}, nil
		},
	}

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	embedding := make([]float32, 768)
	for i := range embedding {
		embedding[i] = 0.5
	}
	insertTestRepoWithEmbedding(t, database, "user/repo", "Content", embedding)

	// Should not panic with nil callback
	result, err := AskQuestion(context.Background(), AskOptions{
		Question:          "What is this repo about?",
		DB:                database,
		EmbeddingProvider: mockEmbed,
		LLMProvider:       mockLLM,
	}, nil)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	if result.Answer != "Answer" {
		t.Errorf("Expected 'Answer', got '%s'", result.Answer)
	}
}

func TestAskQuestion_CaseInsensitiveRepoFilter(t *testing.T) {
	mockEmbed := provider.NewMockEmbeddingProvider()
	mockEmbed.QueryFunc = func(_ context.Context, _ string) []float32 {
		embedding := make([]float32, 768)
		for i := range embedding {
			embedding[i] = 0.5
		}
		return embedding
	}

	mockLLM := &provider.MockLLMProvider{
		NameVal: "mock",
		StreamFunc: func(_ context.Context, _ provider.LLMRequest, onChunk func(string)) (*provider.LLMResult, *provider.LLMError) {
			text := "Filtered answer"
			if onChunk != nil {
				onChunk(text)
			}
			return &provider.LLMResult{Text: text, InputTokens: 10, OutputTokens: 5}, nil
		},
	}

	dbPath := filepath.Join(t.TempDir(), "test.db")
	database, err := db.Open(dbPath, 768)
	if err != nil {
		t.Fatalf("Failed to open database: %v", err)
	}
	defer func() { _ = database.Close() }()

	embedding := make([]float32, 768)
	for i := range embedding {
		embedding[i] = 0.5
	}
	insertTestRepoWithEmbedding(t, database, "User/MyRepo", "Content", embedding)

	// Filter with different case - should still match
	result, err := AskQuestion(context.Background(), AskOptions{
		Question:          "What is this repo about?",
		Repo:              "user/myrepo", // lowercase
		DB:                database,
		EmbeddingProvider: mockEmbed,
		LLMProvider:       mockLLM,
	}, nil)

	if err != nil {
		t.Fatalf("Expected no error, got %v", err)
	}

	// Should have found the repo (case-insensitive match)
	if len(result.Sources) == 0 {
		t.Error("Expected to find sources with case-insensitive match")
	}
}

func TestBuildSourceAttributions_CapsAtFive(t *testing.T) {
	results := []search.SearchResult{
		{RepoFullName: "user/repo1", ChunkType: "readme", Similarity: 0.9},
		{RepoFullName: "user/repo2", ChunkType: "readme", Similarity: 0.8},
		{RepoFullName: "user/repo3", ChunkType: "readme", Similarity: 0.7},
		{RepoFullName: "user/repo4", ChunkType: "readme", Similarity: 0.6},
		{RepoFullName: "user/repo5", ChunkType: "readme", Similarity: 0.5},
		{RepoFullName: "user/repo6", ChunkType: "readme", Similarity: 0.4},
		{RepoFullName: "user/repo7", ChunkType: "readme", Similarity: 0.3},
	}

	sources := buildSourceAttributions(results)

	if len(sources) > 5 {
		t.Errorf("Expected at most 5 sources, got %d", len(sources))
	}
}
